File: oop/04-abstraction/csharp/Account.cs
/*
 * Demonstrates: Abstraction - Abstract Class
 * Concept: Abstract class with both implemented and abstract methods that subclasses must override
 */

namespace FinanceDomain.Abstraction;

public abstract class Account
{
    public string AccountNumber { get; }
    public string AccountHolder { get; }
    protected decimal _balance;

    public decimal Balance => _balance;

    protected Account(string accountNumber, string accountHolder, decimal openingBalance)
    {
        AccountNumber = accountNumber;
        AccountHolder = accountHolder;
        _balance = openingBalance;
    }

    public void Deposit(decimal amount)
    {
        if (amount <= 0)
            throw new ArgumentException("Deposit amount must be greater than zero.");

        _balance += amount;
    }

    public abstract void Withdraw(decimal amount);

    public abstract decimal CalculateMonthlyFees();

    public abstract string GetAccountSummary();

    public void ApplyMonthlyFees()
    {
        decimal fees = CalculateMonthlyFees();
        if (fees > 0)
        {
            _balance -= fees;
        }
    }
}

File: oop/04-abstraction/typescript/account.ts
/*
 * Demonstrates: Abstraction - Abstract Class
 * Concept: Abstract class with both implemented and abstract methods that subclasses must override
 */

export abstract class Account {
    public readonly accountNumber: string;
    public readonly accountHolder: string;
    protected _balance: number;

    constructor(accountNumber: string, accountHolder: string, openingBalance: number) {
        this.accountNumber = accountNumber;
        this.accountHolder = accountHolder;
        this._balance = openingBalance;
    }

    get balance(): number {
        return this._balance;
    }

    deposit(amount: number): void {
        if (amount <= 0) {
            throw new Error("Deposit amount must be greater than zero.");
        }
        this._balance += amount;
    }

    abstract withdraw(amount: number): void;

    abstract calculateMonthlyFees(): number;

    abstract getAccountSummary(): string;

    applyMonthlyFees(): void {
        const fees = this.calculateMonthlyFees();
        if (fees > 0) {
            this._balance -= fees;
        }
    }
}
